File: percolate.m
function [Y,Z]=percolate(T,X,k,p)

%S3DB_PERMIT_CALC percolation of s3db permissions to calculate its effective value at each node
%Syntax: [Y,Z]=percolate(T,X,k)
%   T is a binary transition square matrix and X is the vector array with
%   the permissions. Note that this function works both with migrated and
%   unmigrated permissions. The third input argument indicates the length
%   of the permission state, which will be used in the migration between
%   transitions (inheritances) with the assistance of S3DB_PERMIT_MIGRATE.
%   The 4th input argument allows the use of othe permission codes than the
%   alphabetic order of whatever was used.
%
%   The seccond output argument will reccord the states for all iterations.
%   Consequently, the first column of Z is X and the last is Y.
%
%   The default value for k is 3
%
%   Example:
%
%   The behaviour of the percolation is illustrated for this 5 node
%   interaction
%
%   A-->B, A-->D; B-->C; C-->D; D-->E; E-->A
%
%   which corresponds to a transition matrix T,
%
%   T=[0 1 0 1 0;0 0 1 0 0;0 0 0 1 0;0 0 0 0 1;1 0 0 0 0]
%
%   T =
%
%     0     1     0     1     0
%     0     0     1     0     0
%     0     0     0     1     0
%     0     0     0     0     1
%     1     0     0     0     0
%
%   recalling that the first column of Z is the initial state, X, and the
%   last column is the final, stable, state, Y, analyse these solutions
%
%   X={'n','','','',''}
%
%Z=  'n'    'nnn'    'nnn'
%     ''    'nnn'    'nnn'
%     ''    '---'    'nnn'
%     ''    'nnn'    'nnn'
%     ''    '---'    'nnn'
%
%   X={'yny','yss','','',''}
%
%Z=  'yny'    'yny'    'yny'    'yny'    'ysy'
%    'yss'    'ysy'    'ysy'    'ysy'    'ysy'
%       ''    'yss'    'ysy'    'ysy'    'ysy'
%       ''    'yny'    'ysy'    'ysy'    'ysy'
%       ''    '---'    'yny'    'ysy'    'ysy'
%
%   X={'yNs','ysy','','',''}
%
%Z=  'yNs'    'yNs'    'yNs'    'yNs'    'yNy'
%    'ysy'    'yNy'    'yNy'    'yNy'    'yNy'
%       ''    'ysy'    'yNy'    'yNy'    'yNy'
%       ''    'yNs'    'yNy'    'yNy'    'yNy'
%       ''    '---'    'yNs'    'yNy'    'yNy'
%
%   X={'yNs-','ysy','','',''}
%
%Z=  'yNs-'    'yNs-'    'yNs-'    'yNs-'    'yNy-'
%    'ysy'     'ysy'     'ysy'     'ysy'     'ysy' 
%        ''    'ysy'     'ysy'     'ysy'     'ysy' 
%        ''    '---'     'ysy'     'ysy'     'ysy' 
%        ''    '---'     '---'     'ysy'     'ysy' 
%
%Jonas ALmeida, 31 December 2008

% set defaults
if nargin<3;k=3;end
if size(X,2)>1;X=X';end % to make sure X is processed as a vertical vector
n=length(X); %number of nodes
if size(T,1)~=n|size(T,1)~=size(T,2)
    error(['Transition matrix should be square with side of equal length to the state vector: ',num2str(n),'x',num2str(n)])
end


%calculations
Z=X; %start with current states
T=T==1; %in case T was provided in a non-binary format
% Find which transitions need computing
for g=2:100 %set maximum number of iterations here
    for j=1:n %for each column of the transition matrix
        Xj=migrate(X{j},k,0); %include one's own (no migration needed, it is there already
        for i=[find(T(:,j))]' %for those found to end in the j node
            Xj=[Xj,{migrate(Z{i,g-1},k,1)}]; % migrate permissions into it
        end
        Z(j,g)={s3db_merge(Xj)};
    end
    if strcmp([Z{:,g}],[Z{:,g-1}]); %if there was no change between two generations
        Z(:,end)=[]; % remove last (repeated) column
        break; %stop iteration
    end
end
Y=Z(:,end);

File: percolate.js
// PERCOLATE finds steady state solution for the propagation of s3db operator states
//
//
//
// Syntax: child_state=migrate(parent_state,m,g)
// Description: migrates s3db operator states between a parent and its child
//              entities, as described by Equation 6. The function has two 
//              additional, optional, input arguments to set the state's length,m,
//              and the number of generations, g, i.e. consecutive migrations
//              to be calculated.
//
// Note: console.log commands will report progress back to console if active, 
//       for example, in Firefox, using the firebug addon.
//
// Examples: merge("b,c,C,D") -> "C";  merge("b,cd,Ca,DcA") -> "CdA"
//           merge(["b","cd","Ca","DcA"]) -> "CdA"
//           in the last two examples the console, if active, would show:
//
//              merge("b,c,C,D")="C"
//              merge("-,d,a,c")="d"
//              merge("-,-,-,A")="A"
//
// Jonas Almeida Dec 2009


function percolate(T,S,m) {// EQUATION 7, which is the steady state Solution to Equation 4, below
	if(!T){T=percolate.T}; // transition matrix, T
	if(!S) {S=percolate.S}; // States, S, of model entities E
	if(!m) {if (!percolate.m) {percolate.m=1} ; m=percolate.m}; // number of operators
	S = [S[0]]; // reboot percolation
	var stable = 1;
	for (var i=1; i<100; i++) {
		S[i] = percolate.iterate(T,S[i-1],m);
		// check if stable solution reached
		stable=1;
		for (var j=0; j<S[i].length; j++) {stable=stable*(S[i][j]===S[i-1][j])};
		if (stable) {break};
		// maximum num of iterations
		if (i>10) {break}; 
	}	
    return S;	
}

percolate.iterate = function (T,Sx,m) { // EQUATION 4: Sy = T x Sx , m = number of operators
	if (!m) {m=1}; // default is one operator
	var Sy = new Array;
	for (var i=0; i<Sx.length; i++) {
		var Si = new Array;
		Si[0]=Sx[i]; // the state currently native to the ith entity
		for (var j=0; j<T[i].length; j++) {
			if (T[i][j]===1) {Si[Si.length]=migrate(Sx[j],m)}
		}
		Sy[i]=merge(Si);
	}
	return Sy;
}


percolate.GUI = function (div,T,S) {// create GUI within target div element
	// T is either the transition matrix or the number of entities
	// Identify div
	//console.log('percolateGUI called');
	if (!div) {var div = document.getElementById('percolateDiv')}
	if (typeof(div)=="string") {div = document.getElementById(div)} // in case the div provided is div's id
	// Identify T
	if (parseInt(T)==T) { // if T is an integer specifying the dimensionality of the transition matrix
		var n=parseInt(T);percolate.n=n;T=percolate.createT(n);
	}
	percolate.T=T; // note the transition matrix is being kept as an attribute of the object
	div.innerHTML="<table><tr><td>Transition matrix<br>(Bolean)</td><td>Assigned States<br>(alphabetic indexes)</td><td>Effective States<br>(solved using Eq. 7)</td><td>Percolation of States<br>(iterated Eq. 4)</td></tr><tr><td id=transition></td><td id=states></td><td id=effective_states></td><td id=percolated_states></td></tr>";
	document.getElementById('transition').appendChild(percolate.tableT(T));
	if (!S) {S=percolate.createS()}
	percolate.S=S;
	document.getElementById('states').appendChild(percolate.tableS0(S));
	var bt = document.createElement('input');
	bt.id='runPercolate';
	bt.type='button';
	bt.value="Run";
	bt.onclick=percolate.runGUI;
	div.appendChild(bt);
	return false
}

percolate.runGUI = function () {// run percolation as defined by GUI
	// capture T, S and m
	percolate.m=parseInt(document.getElementById("numOfOperatorsPercolating").value);
	var T = percolate.T;var S = percolate.S;var n = T.length;var Tij=0;
	for (var i=0; i<n; i++) {
		S[0][i]=document.getElementById("S_"+i).value;
		for (var j=0; j<n; j++) {
			Tij = document.getElementById("T_"+i+"_"+j).value;
			if (Tij!=="0") { // make sure T is Bolean: if it is not 0 then it is 1
				Tij=1;
				document.getElementById("T_"+i+"_"+j).value=1;
				T[i][j]=Tij;
			}
		}
	}
	percolate.T=T;
	percolate.S=[S[0]];
	percolate.S=percolate();
	// display percolated states
	percolate.remove_element_by_id("percolatedStatesTable"); // first remove table if it exists
	document.getElementById("percolated_states").appendChild(percolate.tableSs());
	// display effective states
	percolate.remove_element_by_id("effectiveState")
	document.getElementById("effective_states").appendChild(percolate.tableSf());

	return false;
	}

percolate.createT = function (n) {// create a n x n square matriz of zeros
    var T = new Array;
	for (var i=0; i<n; i++) {
		T[i] = new Array;
		for (var j=0; j<n; j++) {
			T[i][j]=0;
		}
	}
	return T;	
}

percolate.s3dbT = function () { // fills S3DB transition matrix in GUI table
	document.getElementById("numOfOperatorsPercolating").value=3;
	document.getElementById('numOfEntities').value=7;
	document.getElementById('T_1_0').value=1;
	document.getElementById('T_2_1').value=1;
	document.getElementById('T_3_1').value=1;
	document.getElementById('T_3_2').value=1;
	document.getElementById('T_3_4').value=1;
	document.getElementById('T_4_2').value=1;
	document.getElementById('T_5_3').value=1;
	document.getElementById('T_5_4').value=1;
	document.getElementById('T_6_0').value=1;
	document.getElementById('T_6_6').value=1;	
}

percolate.createS = function (n) {// create a 1 x n vector of native states (notice teh transposition with regard to regular external product
	if (!n&!!percolate.T) {n=percolate.T.length}
    var S = new Array;
	S[0] = new Array;
	for (var i=0; i<n; i++) {
		S[0][i]="-";
	}	
	return S;	
}


percolate.tableT = function (T) {// create input table for values of T
	if (!T) {T=percolate.T}
    var tb = document.createElement('table');
	var tbody = document.createElement('tbody');
	tb.appendChild(tbody);
	for (var i in T) {
		var tr = document.createElement('tr');
		tbody.appendChild(tr);
		for (var j in T[i]) {
			//console.log(T[i][j])
			var td = document.createElement('td');
			td.innerHTML="<input id='T_"+i+"_"+j+"' size=1 value="+T[i][j]+">";	
			tr.appendChild(td);
		}
		// Native State
	}
	return tb;	
}

percolate.tableS0 = function (S) {// create input table for initial values of S
	if (!S) {S=percolate.S}
    var tb = document.createElement('table');
	var tbody = document.createElement('tbody');
	tb.appendChild(tbody);
	for (var i in S[0]) { // first row contains native state
		var tr = document.createElement('tr');
		tbody.appendChild(tr);
		var td = document.createElement('td');
		td.innerHTML="<input id='S_"+i+"' size=5 value="+S[0][i]+">";	
		tr.appendChild(td);
	}
	return tb;	
}

percolate.tableSf = function (S) {// create input table with final values of S
	if (!S) {S=percolate.S}
    var tb = document.createElement('table');
	tb.id="effectiveState";
	var tbody = document.createElement('tbody');
	tb.appendChild(tbody);
	var n = S.length;
	for (var i in S[n-1]) { // first row contains native state
		var tr = document.createElement('tr');
		tbody.appendChild(tr);
		var td = document.createElement('td');
		td.innerHTML="<input id='Sf_"+i+"' size=5 value="+S[n-1][i]+">";	
		tr.appendChild(td);
	}
	return tb;	
}

percolate.tableSs = function (S) {// create input table for values of T
	if (!S) {S=percolate.S}
    var tb = document.createElement('table');
	//tb.style.borderSpacing=5;
	tb.border=1;
	tb.id = 'percolatedStatesTable';
	var tbody = document.createElement('tbody');
	tb.appendChild(tbody);
	for (var i in S[0]) {
		var tr = document.createElement('tr');
		tbody.appendChild(tr);
		for (var j in S) {
			//console.log(T[i][j])
			var td = document.createElement('td');
			//td.innerHTML="<input id='S_"+i+"_"+j+"'  size=1 value="+S[j][i]+">";
			td.innerHTML=S[j][i];
			tr.appendChild(td);
		}
		// Native State
	}
	return tb;	
}

percolate.remove_element_by_id=function (id) {
	if (document.getElementById(id)) {
		var e = document.getElementById(id);
		e.parentNode.removeChild(e);
	}
	return false;
}

console.log('percolate.js loaded');
